### DISCUSSION TITLE
[Proposal] Overload resolution hint for overloads that have exactly the same priority

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#964] [General]
@ashmind commented on [Thu Oct 01 2015](https://github.com/dotnet/roslyn/issues/5605)

#### Problem

There are cases when there is need for two overloads that perform a same function over different types.
Example:

``` csharp
TValue GetValueOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key);
TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key);
```

This pattern breaks when something belongs to both types, e.g. `d.GetValueOrDefault` would break if `d` is a `Dictionary<,>`, since it implements both interfaces. There is no good workaround other than forcing user to cast, or implementing additional overloads for common types.

Another case (more of an edge case) I had recently was

``` csharp
bool operator==(MyType x, OtherType1 y);
bool operator==(MyType x, OtherType2 y);
```

which breaks on a common case of `x == null` as `null` can belong to either type.
#### Proposal

I feel that in most of those cases it does not matter which overload will be selected for types that match both -- as long as it is actually selected and user does not have to cast.

So the suggestion is to provide a hint to the compiler -- specifying which overload to pick when all of them have exactly the same priority.

To avoid polluting the language, I suggest an attribute, e.g.:

``` csharp
TValue GetValueOrDefault<TKey, TValue>([PrimaryOverload] this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key);
```

``` csharp
bool operator==(MyType x, [PrimaryOverload] OtherType1 y);
```

Not sure what would be the best name -- open to feedback.

---

@MgSam commented on [Thu Oct 01 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-144881052)

What happens in the inevitable situation when multiple methods both specify `PrimaryOverload`? You're back in the same situation as today. Maybe an integer (or double) `OverloadPriority()` attribute works better. Then, like with z-order in graphics, anyone can come in and pick one that's appropriate to give them the behavior they want.

---

@ashmind commented on [Thu Oct 01 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-144881904)

> What happens in the inevitable situation when multiple methods both specify PrimaryOverload? You're back in the same situation as today.

Not quite. Today the API developer has no control over overload selection in cases like `IDictionary` vs `IReadOnlyDictionary`. There is literally no good way for the API to provide that. What I want is to give some control to API developer, making life easier for API users (no cast required).

What you are saying is API developers might mess up their chance at making things right. Well, yes, but it would be a fixable mess up, while right now there is no API-side solution.

---

@ashmind commented on [Thu Oct 01 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-144882476)

> Maybe an integer (or double) OverloadPriority() attribute works better.

I want to avoid this as the suggestion is only for case where overloads have exactly the same priority. Providing a numeric priority would be confusing with regards to other (e.g. type-based) priority rules.

Just to be clear, I don't want a way to prioritize my method over someone else's methods or extension methods. The main goal is to prioritise within a set of methods you own.

---

@MgSam commented on [Fri Oct 02 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145016304)

@ashmind You reference DLL A, which contains `TValue GetValueOrDefault<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key)`. Author A has decorated it with your proposed attribute. 

You reference DLL B, which contains `TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key)`. Author B has also decorated it with your proposed attribute.

Which one gets selected by overload resolution? You're in the same situation you are today. 

Now you could say that this feature only applies to declared in the same assembly- that might be a reasonable compromise.

---

@ashmind commented on [Fri Oct 02 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145137967)

@MgSam You are right, two libraries can conflict. But it's not new -- it's already a conflict even without attributes. It's even likely that two authors with same goals make the same methods (`IDictionary<,>.GetValueOrDefault` in both, for example).

Same as now, these conflicts are resolved by library users who decided to install both, e.g. through imports.

> Now you could say that this feature only applies to declared in the same assembly- that might be a reasonable compromise.

That would only be important if one author used the attribute, and other didn't -- not sure it's worth an extra rule. If both used the attribute, the result of ignoring attribute is same as supporting it -- ambiguity.

---

@gafter commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145578736)

@ashmind Can you please be more specific about what you mean by "priority"? Is this an early tie-breaker or a late tie-breaker? More generally, exactly where in the overload resolution process would this affect the result? Is it intended to apply when neither of two methods are more specific than the other, or when both of two methods are as specific as each other, or in both of these circumstances?

---

@ashmind commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145676700)

@gafter There are two options I considered, either of which will cover the use cases.
I think first option is probably the best, as it is the simplest, but the second one might represent the intent better.

In either option, I am not fond of the attribute name — any suggestions are welcome.

#### Member-level (simplest)

If we say the the attribute goes on the methods themselves, then the syntax would be

``` csharp
[BetterOverload] void M(X x);
void M(Y y);
```

Then to handle it, in C# 5.0 Specification (haven't found 6.0), 7.5.3.2 Better function member, immediately before

> - Otherwise, neither function member is better.

we add
- Otherwise if one member is annotated with BetterOverloadAttribute, and other is not annotated with BetterOverloadAttribute, the annotated one is better.

#### Parameter-level

If we say that attribute goes on the parameters, then the syntax would be

``` csharp
void M([BetterOverload] X x);
void M(Y y);
```

In section 7.5.3.2, we change:

> - Otherwise, if M<sub>P</sub> has more specific parameter types than M<sub>Q</sub>, then M<sub>P</sub> is better than M<sub>Q</sub>. Let { R<sub>1</sub>, R<sub>2</sub>, …, R<sub>N</sub> } and { S<sub>1</sub>, S<sub>2</sub>, …, S<sub>N</sub> } represent the uninstantiated and unexpanded parameter types of M<sub>P</sub> and M<sub>Q</sub>. M<sub>P</sub>’s parameter types are more specific than M<sub>Q</sub>’s if, for each parameter, R<sub>X</sub> is not less specific than S<sub>X</sub>, and, for at least one parameter, R<sub>X</sub> is more specific than S<sub>X</sub>:
>   - A type parameter is less specific than a non-type parameter.
>   - Recursively, a constructed type is more specific than another constructed type (with the same number of type arguments) if at least one type argument is more specific and no type argument is less specific than the corresponding type argument in the other.
>   - An array type is more specific than another array type (with the same number of dimensions) if the element type of the first is more specific than the element type of the second.

to
- Otherwise, if M<sub>P</sub> has more specific parameters than M<sub>Q</sub>, then M<sub>P</sub> is better than M<sub>Q</sub>. Let { R<sub>1</sub>, R<sub>2</sub>, …, R<sub>N</sub> } and { S<sub>1</sub>, S<sub>2</sub>, …, S<sub>N</sub> } represent the  parameters of M<sub>P</sub> and M<sub>Q</sub>, and let the **parameter type** represent uninstantiated and unexpanded parameter type. M<sub>P</sub>’s parameters are more specific than M<sub>Q</sub>’s if each R<sub>X</sub> is not less specific than S<sub>X</sub>, and at least one R<sub>X</sub> is more specific than S<sub>X</sub>:
  - For parameter type, a type parameter is less specific than a non-type parameter.
  - For parameter type, recursively, a constructed type is more specific than another constructed type (with the same number of type arguments) if at least one type argument is more specific and no type argument is less specific than the corresponding type argument in the other.
  - For parameter type, an array type is more specific than another array type (with the same number of dimensions) if the element type of the first is more specific than the element type of the second.
  - For parameter definition, a parameter annotated with BetterOverloadAttribute is more specific than a parameter not annotated with BetterOverloadAttribute.

---

@gafter commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145715681)

@ashmind In both options, you are adding a new bullet under the paragraph

> In case the parameter type sequences {P1, P2, …, PN} and {Q1, Q2, …, QN} are equivalent (i.e. each Pi has an identity conversion to the corresponding Qi), the following tie-breaking rules are applied, in order, to determine the better function member. 

Since P1 is IDictionary and Q1 is IReadOnlyDictionary, the sequences are not equivalent. Therefore neither change would help in the use case that motivated the proposal.

You've also messed up the intended meaning of this section, which was to provide a tie breaker when two _identical_ method signatures are only identical because of instantiation and expansion of the method and enclosing types. None of that applies in your use case.

---

@ashmind commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145723144)

@gafter Hmm, you are right about the equivalence part.

I'm not sure about intended meaning of the whole section though -- seems like section overall defines which function is better for two applicable functions (applicable as per 7.5.3.1). What do you mean by "identical" in that case?

Based on the above, does the following make sense?

#### Member-level

At the very end of 7.5.3.2, add the following:
_If none of the preceding rules identifies M<sub>P</sub> as better than M<sub>Q</sub>, and M<sub>P</sub> is annotated with BetterOverloadAttribute, and M<sub>Q</sub> is not annotated with BetterOverloadAttribute, then M<sub>P</sub> is better than M<sub>Q</sub>._

(That does not go as a list item to the "identity conversion" list, just below it)

---

@ashmind commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145725109)

#### Parameter-level (alternative)

Replace

> Given an argument list A with a set of argument expressions { E<sub>1</sub>, E<sub>2</sub>, ..., E<sub>N</sub> } and two applicable function members M<sub>P</sub> and M<sub>Q</sub> with parameter types { P<sub>1</sub>, P<sub>2</sub>, ..., P<sub>N</sub> } and { Q<sub>1</sub>, Q<sub>2</sub>, ..., Q<sub>N</sub> }, M<sub>P</sub> is defined to be a better function member than M<sub>Q</sub> if
> - for each argument, the implicit conversion from E<sub>X</sub> to Q<sub>X</sub> is not better than the implicit conversion from E<sub>X</sub> to P<sub>X</sub>, and
> - for at least one argument, the conversion from E<sub>X</sub> to P<sub>X</sub> is better than the conversion from E<sub>X</sub> to Q<sub>X</sub>.

With

_Given an argument list A with a set of argument expressions { E<sub>1</sub>, E<sub>2</sub>, ..., E<sub>N</sub> } and two applicable function members M<sub>P</sub> and M<sub>Q</sub> with parameters { P<sub>1</sub>, P<sub>2</sub>, ..., P<sub>N</sub> } and { Q<sub>1</sub>, Q<sub>2</sub>, ..., Q<sub>N</sub> }, M<sub>P</sub> is defined to be a better function member than M<sub>Q</sub> if_
- _for each argument, the implicit conversion from E<sub>X</sub> to type of Q<sub>X</sub> is not better than the implicit conversion from E<sub>X</sub> to type of P<sub>X</sub>, and_
- _for at least one argument, one the following applies:_
  - _the conversion from E<sub>X</sub> to type of P<sub>X</sub> is better than the conversion from E<sub>X</sub> to type of Q<sub>X</sub>_
  - _or, P<sub>X</sub> is annotated with BetterOverloadAttribute and Q<sub>X</sub> is not annotated with BetterOverloadAttribute._

---

@gafter commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145740304)

@ashmind Re "I'm not sure about intended meaning of the whole section though"

That section is for when two overloads that were originally different collide (become identical) because of generic instantiation and/or `params` expansion.

---

@ashmind commented on [Mon Oct 05 2015](https://github.com/dotnet/roslyn/issues/5605#issuecomment-145742273)

@gafter OK I see the confusion, I just understood your "section" term as referring to the whole of 7.5.3.2 and so was confused as it's not just about identical overloads. Now I see it, thanks.

Anyway, does my new spec make sense?

---

@jnm2 commented on [Mon Nov 21 2016](https://github.com/dotnet/roslyn/issues/5605#issuecomment-262048000)

I want this so bad. I have the exact same `IReadOnlyDictionary` vs `IDictionary` extension method pain.

Is it possible to need a tiebreaker between three overloads but at the same time need another tiebreaker between just two of them for a different call site? If so we'll need a priority parameter. (yuck)

---

@ashmind commented on [Mon Apr 10 2017](https://github.com/dotnet/roslyn/issues/5605#issuecomment-293100425)

@gafter Is there an automatic way to migrate this to dotnet/csharplang?

---

@jnm2 commented on [Tue Oct 03 2017](https://github.com/dotnet/roslyn/issues/5605#issuecomment-334011126)

@ashmind Anyone can migrate this with ZenHub or a similar tool. Do you want to, or shall I?

**Comment by Thaina** [comment]
Was having this problem making stackoverflow exception because it calling recursively when I expect it to call underlying function
